File: 0416-Partition Equal Subset Sum/0416-Partition Equal Subset Sum.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int n = nums.size();
        int sum = 0;
        for (int& num : nums) {
            sum += num;
        }
        
        if (sum % 2) {
            return false;
        }
        
        int target = sum / 2;
        vector<bool> dp(1 + target);
        dp[0] = true;
        for (int i = 0; i < n; ++i) {
            for (int j = target; j >= nums[i]; --j) {
                if (dp[j - nums[i]]) {
                    dp[j] = true;
                }
            }
        }
        
        return dp[target];
    }
};

File: 0401-0500/0416-Partition Equal Subset Sum/0416-Partition Equal Subset Sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        total = sum(nums)
        if total & 1:
            return False
        target = total // 2
        dp = [False] * (1 + target)
        dp[0] = True
        for num in nums:
            for j in range(target, num - 1, -1):
                if dp[j - num]:
                    dp[j] = True
        return dp[target]
